### DISCUSSION TITLE
Installation the hub version 3.21.3 partially failed by using cf-remote

### DISCUSSION THREAD

**Discussion by klin938** [cfengine/core#5370] [General]
Hi guys,

I am trying to deploy a CFEngine demo, following the guide: https://docs.cfengine.com/docs/3.21/getting-started-installation.html

The process stuck in the bootstrap the hub itself (that's how I understand it)
```
(cfe_demo) ~$ cf-remote install --hub hub_dev --bootstrap hub_dev

root@50.18.136.51
OS            : Debian 12
Architecture  : x86_64
CFEngine      : Not installed
Policy server : None
Binaries      : dpkg, apt

Copying: '/.cfengine/cf-remote/packages/cfengine-nova-hub_3.21.3-1.debian11_amd64.deb' to 'root@50.18.136.51'
[50.18.136.51]:+ /.cfengine/cf-remote/packages/cfengine-nova-hub_3.21.3-1.debian11_amd64.deb -> cfengine-nova-hub_3.21.3-1.debian11_amd64.deb
Installing: 'cfengine-nova-hub_3.21.3-1.debian11_amd64.deb' on 'root@50.18.136.51'
CFEngine 3.21.3 (Enterprise) was successfully installed on 'root@50.18.136.51'
Bootstrapping: '50.18.136.51' -> '50.18.136.51'
```

Here is what I found from the hub itself:
```
admin@ip-172-31-7-78:~$ systemctl status cf-execd
○ cf-execd.service - CFEngine Execution Scheduler
     Loaded: loaded (/lib/systemd/system/cf-execd.service; enabled; preset: enabled)
     Active: inactive (dead)
  Condition: start condition failed at Wed 2023-11-08 23:37:01 UTC; 11min ago
             └─ ConditionPathExists=/var/cfengine/inputs/promises.cf was not met

Nov 08 23:37:01 ip-172-31-7-78 systemd[1]: cf-execd.service - CFEngine Execution Scheduler was skipped because of an unmet condition check (ConditionPathExists=/var/cfengine/inputs/promises.cf).
admin@ip-172-31-7-78:~$ systemctl status cf-serverd
○ cf-serverd.service - CFEngine Enterprise file server
     Loaded: loaded (/lib/systemd/system/cf-serverd.service; enabled; preset: enabled)
     Active: inactive (dead)
  Condition: start condition failed at Wed 2023-11-08 23:37:01 UTC; 11min ago
             └─ ConditionPathExists=/var/cfengine/inputs/promises.cf was not met

Nov 08 23:37:01 ip-172-31-7-78 systemd[1]: cf-serverd.service - CFEngine Enterprise file server was skipped because of an unmet condition check (ConditionPathExists=/var/cfengine/inputs/promises.cf).
```
The contents of inputs directory:
```
admin@ip-172-31-7-78:~$ sudo ls /var/cfengine/inputs/
cf_promises_release_id  failsafe.cf
```

I am wondering if I have missed any important steps? 

Cheesr
D

**Comment by nickanderson** [comment]
It says it bootstrapped,  but inputs doesn't look like it should.

Off the top of my head I don't know where if cf-remote logs see that bootstrap output. 

I would just try bootstrapping again, cf-agent - - bootstrap ip on the host as root.

**Reply by klin938** [reply]
Thanks Nick,

I tried bootstrap it manually:

```
admin@ip-172-31-7-78:/var/cfengine$ sudo /var/cfengine/bin/cf-agent --bootstrap 50.18.136.51
  notice: Bootstrap mode: implicitly trusting server, use --trust-server=no if server trust is already established
   error: No suitable server found for '/var/cfengine/inputs'
   error: Promise belongs to bundle 'failsafe_cfe_internal_update' in file '/var/cfengine/inputs/failsafe.cf' near line 121
   error: Errors encountered when actuating files promise '/var/cfengine/inputs'
   error: No suitable server found for '/var/cfengine/modules'
   error: Promise belongs to bundle 'failsafe_cfe_internal_update' in file '/var/cfengine/inputs/failsafe.cf' near line 130
   error: Errors encountered when actuating files promise '/var/cfengine/modules'
   error: No suitable server found for '/var/cfengine/inputs'
   error: Promise belongs to bundle 'failsafe_cfe_internal_update' in file '/var/cfengine/inputs/failsafe.cf' near line 144
   error: Comment is 'If we failed to fetch policy we try again using
                    the legacy default in case we are fetching policy
                    from a hub that is not serving mastefiles via a
                    shortcut.'
   error: Errors encountered when actuating files promise '/var/cfengine/inputs'
   error: Method 'failsafe_cfe_internal_update' failed in some repairs
R: Bootstrapping from host '50.18.136.51' via built-in policy '/var/cfengine/inputs/failsafe.cf'
R: This autonomous node assumes the role of voluntary client
   error: Bootstrapping failed, no input file at '/var/cfengine/inputs/promises.cf' after bootstrap
```
Should I use localhost or 127.0.0.1 when bootstraping the hub itself?

Cheers
D

**Reply by klin938** [reply]
My hub is on a AWS instance, it turned out that the bootstrap process detected its private IP which is associated with the NIC, not the public IP. So using public IP caused the error.

**Reply by larsewi** [reply]
Looks like you might be missing the [masterfiles](https://github.com/cfengine/masterfiles) policy framework. Maybe you've installed a non-hub package on that host? I believe hub packages should include the masterfiles policy framework, if I'm not mistaking. Maybe @nickanderson can confirm if this is true?

**Reply by larsewi** [reply]
Ah, I see you fixed it already 👍

**Comment by nickanderson** [comment]
> My hub is on a AWS instance, it turned out that the bootstrap process detected its private IP which is associated with the NIC, not the public IP. So using public IP caused the error.

Yes. When you bootstrap the hub to itself, you should bootstrap to an IP (or hostname that resolves to) an interface attached to the host. So, looking at your console output, that would be 172.31.7.78 (`cf-agent --bootstrap 172.31.7.78`).

And with `cf-remote` then, it would be something like: `cf-remote install --hub hub_dev --bootstrap 172.31.7.78`

Glad you got it working.

**Reply by 0lmi** [reply]
I encountered the same issue simply following the installation guide [1]. It seems that the guide works only when the Hub is installed from a host within the same private network as the Hub. When installing the Hub from a desktop over the internet, with the Hub in a cloud, the process uses a public IP/FQDN, which causes the bootstrap to fail.

[1] https://docs.cfengine.com/docs/3.24/getting-started-installation.html#install-cfengine
